mgr/prometheus: provide RBD stats via osd dynamic perf counters

trociny opened the pull request:
Signed-off-by: Mykola Golub <mgolub@suse.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

trociny commented:
Example:
```
% ceph osd pool create rbd 32
% rbd pool init
% rbd create test -s 100M
% ceph mgr module enable prometheus
% ceph config set mgr mgr/prometheus/rbd_stats_pools "rbd"

% wget -q -O - http://127.0.0.1:9283/metrics | grep ceph_rbd_
# HELP ceph_rbd_read_bytes_sum RBD image bytes read Total
# TYPE ceph_rbd_read_bytes_sum counter
ceph_rbd_read_bytes_sum{pool="rbd",image="test"} 0.0
# HELP ceph_rbd_write_ops RBD image writes count
# TYPE ceph_rbd_write_ops counter
ceph_rbd_write_ops{pool="rbd",image="test"} 0.0
# HELP ceph_rbd_read_bytes_count RBD image bytes read Count
# TYPE ceph_rbd_read_bytes_count counter
ceph_rbd_read_bytes_count{pool="rbd",image="test"} 0.0
# HELP ceph_rbd_write_bytes_sum RBD image bytes written Total
# TYPE ceph_rbd_write_bytes_sum counter
ceph_rbd_write_bytes_sum{pool="rbd",image="test"} 0.0
# HELP ceph_rbd_read_ops RBD image reads count
# TYPE ceph_rbd_read_ops counter
ceph_rbd_read_ops{pool="rbd",image="test"} 0.0
# HELP ceph_rbd_write_bytes_count RBD image bytes written Count
# TYPE ceph_rbd_write_bytes_count counter
ceph_rbd_write_bytes_count{pool="rbd",image="test"} 0.0
# HELP ceph_rbd_read_latency_sum RBD image reads latency (msec) Total
# TYPE ceph_rbd_read_latency_sum counter
ceph_rbd_read_latency_sum{pool="rbd",image="test"} 0.0
# HELP ceph_rbd_read_latency_count RBD image reads latency (msec) Count
# TYPE ceph_rbd_read_latency_count counter
ceph_rbd_read_latency_count{pool="rbd",image="test"} 0.0
# HELP ceph_rbd_write_latency_count RBD image writes latency (msec) Count
# TYPE ceph_rbd_write_latency_count counter
ceph_rbd_write_latency_count{pool="rbd",image="test"} 0.0
# HELP ceph_rbd_write_latency_sum RBD image writes latency (msec) Total
# TYPE ceph_rbd_write_latency_sum counter
ceph_rbd_write_latency_sum{pool="rbd",image="test"} 0.0

% rbd bench --rbd-cache=false --io-type write --io-size 1M --io-total 100M test  
bench  type write io_size 1048576 io_threads 16 bytes 104857600 pattern sequential
  SEC       OPS   OPS/SEC   BYTES/SEC
elapsed:     1  ops:      100  ops/sec:    99.60  bytes/sec: 104438650.41
% rbd bench --rbd-cache=false --io-type read --io-size 1M --io-total 100M test
bench  type read io_size 1048576 io_threads 16 bytes 104857600 pattern sequential
  SEC       OPS   OPS/SEC   BYTES/SEC
elapsed:     0  ops:      100  ops/sec:  1136.35  bytes/sec: 1191550055.40

% wget -q -O - http://127.0.0.1:9283/metrics |grep 'pool="rbd",image="test"' | sort
ceph_rbd_read_bytes_count{pool="rbd",image="test"} 100.0
ceph_rbd_read_bytes_sum{pool="rbd",image="test"} 104860000.0
ceph_rbd_read_latency_count{pool="rbd",image="test"} 100.0
ceph_rbd_read_latency_sum{pool="rbd",image="test"} 745414124.0
ceph_rbd_read_ops{pool="rbd",image="test"} 100.0
ceph_rbd_write_bytes_count{pool="rbd",image="test"} 100.0
ceph_rbd_write_bytes_sum{pool="rbd",image="test"} 104857600.0
ceph_rbd_write_latency_count{pool="rbd",image="test"} 100.0
ceph_rbd_write_latency_sum{pool="rbd",image="test"} 5097196535.0
ceph_rbd_write_ops{pool="rbd",image="test"} 100.0
```

dillaman commented on src/pybind/mgr/prometheus/module.py:
```diff
@@ -505,6 +508,146 @@ def get_num_objects(self):
+
+        res = self.get_osd_perf_counters(self.rbd_stats['query_id'])
+        for c in res['counters']:
+            if c['k'][1][1]:
```
I would have thought that the pool id would always be in ```c['k'][0][0]``` since it's the first listed sub-key above? Of course, if that's the case, I guess I don't understand why the image id would be in ```c['k'][1][2]``` below.

trociny replied:
Note, the regex is `'^(?:rbd|journal)_data\.(?:([0-9]+)\.)?([^.]+)\.'`. Consider an object "rbd_data.3.imageid.000xyz". The object will be in the data pool, so the pool id from "pool id" subkey (`c['k'][0][0]`) will match the data pool id, but to construct "pool/image" spec we need the metadata pool id, which can be extracted from the object name and is 3 for this example.

BTW for osd_perf_query module we have a simplified rbd query regex, which does not work for data pool and journal objects. I have a patch for it, but have not pushed it yet as it depends on "limit" PR to be  merged first.

dillaman replied:
Ah -- missed the pool id optional in the reg ex. Can you add one-line comment here to clarify the intention?

dillaman commented on src/pybind/mgr/prometheus/module.py:
```diff
@@ -505,6 +508,146 @@ def get_num_objects(self):
+                    pool = self.rbd_stats['pools'][pool_id]
+                    pool['name'] = pool_name
+                    images = {}
+                    for image_meta in RBD().list2(ioctx):
```
It's going to look-up all the images in the pools on each iteration of the prometheus status gathering? It would be great if this could be caches and only refreshed as needed (i.e. unknown image id in the results)

trociny replied:
Not exactly on each iteration -- if it is gathered more frequently than `scrape_interval` then just a cached result is returned, but `scrape_interval` is small by default (5 sec). So yes, I think it will usually fired on every prometheus gathering.

I think I can't just refresh only  if an unknown image id is found in the result. The module reports stats for all images, including inactive. So if an image is added but not used yet, the stats will not be produced. And if an image is removed the module will continue to return its stats.

I was planning to improve this later, but I can do in this PR.  I think I will add an additional param `rbd_stats_pool_refresh_interval`, setting it to 5 minutes, so it will not refresh on every gathering. And will also add force refresh when an unknow image id found in the result.

dillaman replied:
Sounds good

dillaman commented on src/pybind/mgr/prometheus/module.py:
```diff
@@ -156,6 +157,7 @@ class Module(MgrModule):
             {'name': 'server_addr'},
             {'name': 'server_port'},
             {'name': 'scrape_interval'},
+            {'name': 'rbd_stats_pools'},
```
Nit: can you add a "max result" limit optional and default it to something semi-sane (like top 20 or 50)? If someone overrides it, at least it's not configured out-of-the-box to grab too much data.

trociny replied:
I am not sure here. I think I would agree to have some safety limit, just in case it will really produce too much in some deployments (similarly how we have limits like `osd_max_object_size`), but it should be much  higher then 50. Considering that one entry size would be about 80 bytes, I think we could set the limit to 16K (about 1M per message) or even 64K (about 5M per message). I suppose the installations that may hit such a limit (i.e having about 100K active images?) can afford report messages of such a size?

Setting it just to 20 or 50 does not make much sense to me, because it would mean that almost everyone would hit it, and would have incomplete stats, difficult to interpret.

dillaman replied:
Well, it would provide the top N image stats per OSD (note that I still want to use weighted probability to select that top N-ish results at the OSD side in a future PR), not only N images in the result.   I'd be interested to see a test where Prometheus was ingesting 5MiB of stats every 5 seconds (or so) and see if it can keep up with the load. That's why I think it's safer to start small.

trociny replied:
Exactly, if you expect the prometheus will have problems earlier with such amount of data, why we need to limit osds here? It will not help for prometheus (because it still will receive stats for all images in pool), but in most cases with such default settings (top 20 or 50) the stats will be inaccurate. And it will be not so easy for a user to realize what is wrong here.

By default it is disabled. Users should explicitly specify the pool(s) (or just a namespace, I will add the support for this later) they want to monitor. If it causes an issue they can just disable it. Having a high enough limit (16K) is probably good as a safety guard.

I will try to to stress test this (and prometheus) with a large number of images.

dillaman replied:
Sorry -- it was two separate thoughts in there (noting that limiting the stats would be top-N per OSD *and* the query if Prometheus can really handle a large ingestion of data like that).

trociny commented:
Updated: the described above pool refresh logic added and the nits fixed. It is still missing the limit though, because it depends on #25151. And I still have plans to make some measurements to see what a limit could be sane.

trociny commented:
> With the refresh interval, what happens if an image is removed but the module doesn't refresh for another 5 minutes? Is old data exported?

yes, old data is exported, but I don't think it is a big deal.

> Would it make sense to think about adding a current images cache plus notification piece to the mgr? I'm rather ignorant regarding rbd but it works like this for osd maps already (mgr gets notified on osd map changes). This is a bigger pice of work (so doesn't concern this PR) but iiuc this would simplify this code quite a bit and eliminate the "no data" period for new images.

Technically "notification update" implementation for rbd pools is possible (when a mgr is listening on rbd_directory object in every rbd pool, and librbd sends a  notification to the object when adding/removing an image). We already have something like this for rbd-mirror and mirroring images. Not sure though if we really need this here. And it will not work properly when an image is created/removed by an older librbd client.

Having this (up to date rbd images map) in mgr (so it could be shared by different modules) may have advantages, because I suppose when `rbd top`  is implemented it will need to solve the same problem -- map image ids to image names (though it might do this on the client side and probably in a different way). And probably dashboard could benefit from this too.

trociny commented:
repushed: unrelated local changes leaked in the previous push

trociny commented:
@dillaman There is a complication to add limit param to the query: what counter should be used for "order_by". I suppose it should be "bytes" (read + writes), but the query ('performance_counter_descriptors') currently does not contain it. I am not very happy to add it just for this purposes -- I still think that normally the limit should not be reached (otherwise the stats will be incorrect) and an unused counter will just be increasing payload.

trociny commented:
updated after testing:
- don't use `pool['images'].items()` when iterating (a potentially big list of) images, because it will return a list in python2;
- fixed `del self.rbd_stats['pools'][pool_id]` for python3 (RuntimeError: dictionary changed size during iteration).
- cleanup: `.keys()` is not necessary.

The results of stress testing on a local dev server, when running a vstart cluster with 20 osds, and a pool with 10,000 RBD images:
- I was able to generate load with about 750 entries in osd_perf_starts reports.
- it required about up to 3 sec for the prometheus mgr module to generate stats (1.5 sec to scan the pool when needed + 1.5 seconds to actually generate stats).
- the prometheus ingested the stats quite quickly, up to 3.x sec when the pool was refreshed and 1.8 sec when it was not. So it looked mostly stats generation time on the mgr side.

So it looks like 10,000 images not a problem for the prometheus at all. I have not noticed any related issues on osd side.
